src/ast/values.ts

```diff
@@ -12,3 +32,229 @@ export const UNKNOWN_EXPRESSION: ExpressionEntity = {
+				const subPath = path[0];
+				return isUnknownKey(subPath)
+					|| !literalMembers[type][subPath]
+					|| predicateFunction(options)(literalMembers[type][subPath].returns);
+			}
+			return true;
+		},
+		toString: () => '[[UNKNOWN STRING]]'
+	};
+}
+
+export const UNKNOWN_LITERALS = {
+	boolean: createUnknownLiteral('boolean'),
+	number: createUnknownLiteral('number'),
+	string: createUnknownLiteral('string')
+};
+const returnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: null } };
+const callsArgReturnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: [0] } };
+const returnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: null } };
+const callsArgReturnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: [0] } };
+const returnsString: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.string, callsArgs: null } };
+
+export const UNKNOWN_OBJECT_EXPRESSION: ExpressionEntity = {
+	reassignPath: () => {},
+	forEachReturnExpressionWhenCalledAtPath: () => {},
+	getValue: () => UNKNOWN_VALUE,
+	hasEffectsWhenAccessedAtPath: path => path.length > 1,
+	hasEffectsWhenAssignedAtPath: path => path.length > 1,
+	hasEffectsWhenCalledAtPath: path => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath) || !objectMembers[subPath];
+		}
+		return true;
+	},
+	someReturnExpressionWhenCalledAtPath: (
+		path: ObjectPath,
+		_callOptions: CallOptions,
+		predicateFunction: SomeReturnExpressionCallback,
+		options: ExecutionPathOptions
+	) => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath)
+				|| !objectMembers[subPath]
+				|| predicateFunction(options)(objectMembers[subPath].returns);
+		}
+		return true;
+	},
+	toString: () => '[[UNKNOWN OBJECT]]'
+};
+
+export const arrayMembers: MemberDescriptions =
+	assembleMemberDescriptions({
+		concat: returnsArray,
+		every: callsArgReturnsBoolean,
+		filter: callsArgReturnsArray,
+		find: callsArgReturnsUnknown,
+		findIndex: callsArgReturnsNumber,
+		forEach: callsArgReturnsUnknown,
+		includes: returnsBoolean,
+		indexOf: returnsNumber,
+		join: returnsString,
+		lastIndexOf: returnsNumber,
+		map: callsArgReturnsArray,
+		reduce: callsArgReturnsUnknown,
+		reduceRight: callsArgReturnsUnknown,
+		slice: returnsArray,
+		some: callsArgReturnsBoolean,
+		toLocaleString: returnsString,
+		toString: returnsString
+	});
+
+const literalMembers: { [key: string]: MemberDescriptions } = {
+	boolean: assembleMemberDescriptions({
+		toString: returnsString,
+		valueOf: returnsBoolean
+	}),
+	number: assembleMemberDescriptions({
+		toExponential: returnsString,
+		toFixed: returnsString,
+		toLocaleString: returnsString,
+		toPrecision: returnsString,
+		toString: returnsString,
+		valueOf: returnsNumber
+	}),
+	string: assembleMemberDescriptions({
+		charAt: returnsString,
```

Comment:
I can't believe you skipped `string.blink()` 😜

Reply:
I KNOW. Especially since I just checked that even Node supports it! Should I change that?

Reply:
I would personally be deeply upset if my codebase wasn't able to tree-shake my pure blink-rendering code.